src/threadpool.c

```diff
@@ -243,6 +260,7 @@ void uv__work_done(uv_async_t* handle) {
     w = container_of(q, struct uv__work, wq);
     err = (w->work == uv__cancelled) ? UV_ECANCELED : 0;
     w->done(w, err);
+    report(DONE, 1);
```

Comment:
This incurs the overhead of locking and unlocking the mutex for every work item, even when there are no observers.

I'd cache the result of `QUEUE_EMPTY(&stats)` in the critical section a few lines up and only call `report()` if it's not empty.

That would introduce a small when a done callback calls `uv_queue_stats_start()` but that seems like an acceptable trade-off to me; it just needs to be documented.

As well, that would let you drop the second `report()` parameter because you can simply make it a rule that it should only be called with the lock held.

(We should review whether it's always safe to drop and reacquire the lock when invoking the callback though.)